### ISSUE TITLE
Phone not detected when connected to the internet with a wired connection

### ISSUE THREAD

**Issue by Aleksandar204** [GSConnect/gnome-shell-extension-gsconnect#1720]
### Describe the bug

When using my laptop connected with a dock, gsconnect cannot see my phone on the network. Turning off wired internet makes gsconnect recognise the phone over wifi every time. Firewall is configured correctly, and even on the wired connection I am able to see that a connection to the phone is possible (`netcat -z -v 192.168.1.20 1714-1764` displays 'Connection succeeded!' for port 1716). 

### Steps to reproduce

1. Pair phone and laptop over wifi
2. Disable gsconnect or turn off laptop
3. Connect to wired ethernet (leaving wifi on or off doesn't matter, it does not work ever)
4. Laptop does not connect to phone anymore, just says disconnected

### GSConnect version

56

### Installed from

Extension Manager

### GNOME Shell version

45.2

### Linux distribution/release

Fedora 39

### Paired device(s)

Oneplus 7 Pro

### KDE Connect app version

1.29.0

**Comment by genodeftest** [comment]
Same issue here (same software versions, but with Fairphone 4 instead of OnePlus 7 Pro).

# Workaround:
1. On the PC, enable wifi
2. On the PC, disable wired connection

After the PC only has had wifi for a few seconds, GSConnect reconnects to KDEConnect on smartphone. It continues to work after I first enable wired connection, then (after a few seconds) disable wifi.

In other words, it seems like that only for initial reconnection, both devices need to be on the same wifi, but as soon as the connection has been established (and until it gets interrupted), connection keeps working even if both devices are on separate (but connected) networks.

**Comment by ferdnyc** [comment]
@Aleksandar204 

> Firewall is configured correctly, and even on the wired connection I am able to see that a connection to the phone is possible (netcat -z -v 192.168.1.20 1714-1764 displays 'Connection succeeded!' for port 1716).

One thing worth double-checking, are you sure the firewall is correctly configured **_for every interface_**? firewalld manages its profiles (or "Zones") per-interface. Each one has a "Default Zone" (you can see them listed in `firewall-config`), and unless they're the same they're going to have different configs.

A bad firewall config wouldn't prevent you from reaching the phone, but it would prevent the phone from reaching you. The Android app "Fing - Network tools" can do open-port scans. (I'm sure there are others.) It's a little annoying, because you have to sit through a full scan of _all_ its known ports, but it does work. (Actually it _won't_ work to find KDEConnect ports, unless you add them to its known services list. There's a  "SETTINGS" link in the upper-right corner of the portscan screen for that. Adding just 1716 and 1739 is sufficient to confirm that GSConnect is reachable from the phone, in different situations.

@genodeftest 

> 1. On the PC, enable wifi
> 1. On the PC, disable wired connection
>
> After the PC only has had wifi for a few seconds, GSConnect reconnects to KDEConnect on smartphone. It continues to work after I first enable wired connection, then (after a few seconds) disable wifi.
>
> In other words, it seems like that only for initial reconnection, both devices need to be on the same wifi, but as soon as the connection has been established (and until it gets interrupted), connection keeps working even if both devices are on separate (but connected) networks.

Hmmm. I use GSConnect exclusively over a wired PC connection, to the wifi router that my phone is also connected to... but the difference is, my desktop doesn't have WiFi at all.

Presumably, changing network interfaces changes your IP address on the GSConnect side of things. (I assume @Aleksandar204's docking/undocking the laptop does as well, since the dock wouldn't likely share an IP or MAC with the laptop's WiFi.) So it's especially weird that the connection would stay up when switching networks.

Beyond confirming in `firewall-config` that all involved interfaces have the same firewall settings, a look at the [Generate Support Log](https://github.com/GSConnect/gnome-shell-extension-gsconnect/wiki/Help#generate-support-log) info from GSConnect while you're switching networks _might_ be helpful. (You'd want to start it capturing before anything is done, and running until just after the issue is fully reproduced.) I say "might" because GSConnect's network logging isn't the _best_ -- it logs _activity_ fine when you are connected, but when you **can't** connect, generally it just logs nothing.

More informative might be to also have `nmcli m` running over the same timeframe, to see what NetworkManager is doing at various points in the process.

**Comment by genodeftest** [comment]
> > It continues to work after I first enable wired connection, then (after a few seconds) disable wifi.

This does not seem to be true any more.

> Presumably, changing network interfaces changes your IP address on the GSConnect side of things.

It does, at least for me. My desktop PC has two different IP addresses when connected via both Ethernet and Wifi, and it has different IP addresses when connected via Ethernet vs. when connected via Wifi.

> I assume @Aleksandar204's docking/undocking the laptop does as well, since the dock wouldn't likely share an IP or MAC with the laptop's WiFi.

Some combinations of laptop+dock are able to pass through the MAC address, so this might also apply to the IP address. This can be enabled/disabled in the BIOS setup, e.g. in a Dell latitude, where it is specified to work with some docks. (this is not relevant for my setup, but it might apply to @Aleksandar204's setup)

_____
> A bad firewall config wouldn't prevent you from reaching the phone, but it would prevent the phone from reaching you. The Android app "Fing - Network tools" can do open-port scans.

My system runs firewalld. How would I find this out? Do a full port scan of both devices (from Smartphone to PC and vice versa) in both scenarios (when connection works fine vs. when connection does not work)?

_____
Vague idea / guess: When enabling/disabling a network interface on the PC, is it possible that this triggers some broadcast announce message, e.g. on DHCP? And that under certain conditions, the KDEconnect app doesn't "_know_" the PC until this happens?

If that were true, the issue should happen only if the Smartphone joined the network/wifi after the PC (because then it would be notified as soon as the PC joined). But the connection should immediately work if the PC joins after the Smartphone (because when joining the network, the broadcast message is sent).

**Comment by Aleksandar204** [comment]
> One thing worth double-checking, are you sure the firewall is correctly configured for every interface? firewalld manages its profiles (or "Zones") per-interface. Each one has a "Default Zone" (you can see them listed in firewall-config), and unless they're the same they're going to have different configs.

Yes, I have checked it using the gui, and also tested it before with the firewall completely disabled. 

> Presumably, changing network interfaces changes your IP address on the GSConnect side of things. (I assume @Aleksandar204's docking/undocking the laptop does as well, since the dock wouldn't likely share an IP or MAC with the laptop's WiFi.) So it's especially weird that the connection would stay up when switching networks.

My current setup is a Thinkpad T14 G3 with the Lenovo Thinkpad USB C dock (model 40AY I think). It does support MAC address passthrough, but only from the built in LAN port and it doesn't mess with the wireless. Both IPs are left to the routers DHCP server to decide.

> A bad firewall config wouldn't prevent you from reaching the phone, but it would prevent the phone from reaching you. The Android app "Fing - Network tools" can do open-port scans. (I'm sure there are others.) It's a little annoying, because you have to sit through a full scan of all its known ports, but it does work. (Actually it won't work to find KDEConnect ports, unless you add them to its known services list. There's a "SETTINGS" link in the upper-right corner of the portscan screen for that. Adding just 1716 and 1739 is sufficient to confirm that GSConnect is reachable from the phone, in different situations.

Can't test this currently as I am currently not at home for the holidays, but I believe that this was correctly done since I often had ports opened as i had a service running on other ports and I often needed to access it though my phone. Doing this scan on the wifi I'm currently connected to (not my home network) results in port 1716 being open. All interfaces are using the same default zone so idk if it would be any different on my network and the dock.

> Beyond confirming in firewall-config that all involved interfaces have the same firewall settings, a look at the [Generate Support Log](https://github.com/GSConnect/gnome-shell-extension-gsconnect/wiki/Help#generate-support-log) info from GSConnect while you're switching networks might be helpful. (You'd want to start it capturing before anything is done, and running until just after the issue is fully reproduced.) I say "might" because GSConnect's network logging isn't the best -- it logs activity fine when you are connected, but when you can't connect, generally it just logs nothing.
More informative might be to also have nmcli m running over the same timeframe, to see what NetworkManager is doing at various points in the process.

Will try this once  I return home, or sooner if I find an ethernet cable since the LAN port has the same behavior as the dock.

P.S. What exactly does the UDP discovery option do in the KDE connect settings? Shouldn't the broadcast already make the devices find each other, or I'm missing something important here?

**Comment by genodeftest** [comment]
I have not seen this issue in a few weeks now. I guess some update in some tool fixed this issue. Can anyone else confirm this?

(I am using Fedora 39 on my PC, in case that is relevant)

**Comment by Aleksandar204** [comment]
This issue has magically returned for me. Went through testing a few thing and in the end I think the issue is in the extension itself, not my device. Tcpdump for port 5355 shows that my phone advetises kdeconnect as it should, `avahi-browse -a` also shows kdeconnect from my phone reaching the laptop over the enp5s0f4u1u1 port. I triple checked the firewall, so no issues there.

In the end i installed kdeconnect instead of gsconnect and magically it started working. I'm happy to test anything if necessary, but for now I'll stick with kdeconnect

My current system is the same thinkpad+dock combo but now with fedora 43
